app/database/build_store.py

```diff
@@ -0,0 +1,307 @@
+                    build_id=build_id,
+                    failed_artifact_directory=directory
+                )
+                session.add(failed_artifact_directory)
+
+        # FailedSubjobAtomPairs
+        if build._build_artifact is not None:
+            for subjob_id, atom_id in build._build_artifact.get_failed_subjob_and_atom_ids():
+                failed_subjob_and_atom_ids = FailedSubjobAtomPairsSchema(
+                    build_id=build_id,
+                    subjob_id=subjob_id,
+                    atom_id=atom_id
+                )
+                session.add(failed_subjob_and_atom_ids)
+
+        # Subjobs
+        subjobs = build._all_subjobs_by_id
+        for subjob_id in subjobs:
+            subjob = build._all_subjobs_by_id[subjob_id]
+            subjob_schema = SubjobsSchema(
+                subjob_id=subjob_id,
+                build_id=build_id,
+                completed=subjob.completed
+            )
+            session.add(subjob_schema)
+
+            # Atoms
+            for atom in subjob._atoms:
+                atom_schema = AtomsSchema(
+                    atom_id=atom.id,
+                    build_id=build_id,
+                    subjob_id=subjob_id,
+                    command_string=atom.command_string,
+                    expected_time=atom.expected_time,
+                    actual_time=atom.actual_time,
+                    exit_code=atom.exit_code,
+                    state=atom.state,
+                )
+                session.add(atom_schema)
+
+        session.commit()
+        return build_id
+
+    @classmethod
+    def _reconstruct_build(cls, build_id, allow_incompleted_builds=False) -> Build:
+        """
+        Given a build_id, fetch all the stored information from the database to reconstruct
+        a Build object to represent that build.
+        :param build_id: The id of the build to recreate.
+        """
+        session = Connection.get()
+
+        q_build = session.query(BuildSchema).filter(BuildSchema.build_id == build_id).first()
+        q_failed_artifact_directories = session.query(FailedArtifactDirectoriesSchema).filter(FailedArtifactDirectoriesSchema.build_id == build_id).all()
+        q_failed_subjob_atom_pairs = session.query(FailedSubjobAtomPairsSchema).filter(FailedSubjobAtomPairsSchema.build_id == build_id).all()
+        q_build_atoms = session.query(AtomsSchema).filter(AtomsSchema.build_id == build_id).all()
+        q_build_subjobs = session.query(SubjobsSchema).filter(SubjobsSchema.build_id == build_id).all()
+
+        # If a query returns None, then we know the build wasn't found in the database
+        if not q_build:
+            return None
+
+        # Build wasn't completed when we stored it
+        if not allow_incompleted_builds and not bool(int(q_build.completed)):
+            raise IncompleteBuild('Cannot load build (id: {}) because it was never completed.'.format(build_id))
```

Comment:
Again, I would rather have this behavior:
- During shutdown any builds currently in progress are failed with a message like "Master shut down while build was executing."
- Those builds enter the error state. They're loaded just like any other finished build after the master restarts.
- We don't add any code right now around supporting or preserving in-progress builds.

In my mind, attempting to preserve the state of in-progress builds when the master restarts adds a bunch of complexity with no benefit. Maybe in the future we'll decide to support resuming builds after the master restarts, but the changes needed for that are much more extensive than even what you have here. I'm also not convinced that we'll ever do that -- it would be a complicated feature and the benefit we'd get out of it is not very significant.

Reply:
> During shutdown any builds currently in progress are failed with a message like "Master shut down while build was executing."

Could you point to an example of failing a build with an error message?

Reply:
https://github.com/box/ClusterRunner/blob/master@%7B2017-08-24%7D/app/master/build_request_handler.py#L102

Just ignore my #WIP notes. I left a mess. :)